src/add_toppra_time_parameterization.cpp

```diff
@@ -0,0 +1,238 @@
+    const char* ModelUrl = "package://drake_models/franka_description/"
+                           "urdf/panda_arm.urdf";
+    const std::string urdf = PackageMap{}.ResolveUrl(ModelUrl);
+    Parser(&plant, &scene_graph).AddModels(urdf);
+    plant.WeldFrames(plant.world_frame(), plant.GetFrameByName("panda_link0"));
+
+    // for now finalize plant here
+    plant.Finalize();
+
+    diagram_ = builder->Build();
+    diagram_context_ = diagram_->CreateDefaultContext();
+  }
+
+  [[nodiscard]] std::string getDescription() const override
+  {
+    return std::string("AddToppraTimeParameterization");
+  }
+
+  void adapt(const planning_scene::PlanningSceneConstPtr& planning_scene,
+             const planning_interface::MotionPlanRequest& /* req */,
+             planning_interface::MotionPlanResponse& res) const override
+  {
+    // Check if res contains a path
+    if (!res.trajectory)
+    {
+      RCLCPP_ERROR(getLogger(),
+                   "Cannot apply response adapter '%s' because MotionPlanResponse does not contain a path.",
+                   getDescription().c_str());
+      res.error_code = moveit::core::MoveItErrorCode::INVALID_MOTION_PLAN;
+      return;
+    }
+
+    // Get joint model group
+    const moveit::core::JointModelGroup* joint_model_group = res.trajectory->getGroup();
+    if (!joint_model_group)
+    {
+      RCLCPP_ERROR(getLogger(), "It looks like the pipeline did not set the group the plan was computed for");
+      res.error_code = moveit::core::MoveItErrorCode::FAILURE;
+      return;
+    }
+
+    /////////////////////////////////////
+    // Update plan from planning scene //
+    /////////////////////////////////////
+    auto& plant = dynamic_cast<const MultibodyPlant<double>&>(diagram_->GetSubsystemByName("plant"));
+    auto& plant_context = diagram_->GetMutableSubsystemContext(plant, diagram_context_.get());
+    const auto& current_state = planning_scene->getCurrentState();
+    Eigen::VectorXd q_pos = Eigen::VectorXd::Zero(joint_model_group->getActiveVariableCount());
+    Eigen::VectorXd q_vel = Eigen::VectorXd::Zero(joint_model_group->getActiveVariableCount());
+    Eigen::VectorXd q = Eigen::VectorXd::Zero(2 * joint_model_group->getActiveVariableCount());
+
+    current_state.copyJointGroupPositions(joint_model_group, q_pos);
+    current_state.copyJointGroupVelocities(joint_model_group, q_vel);
+    q << q_pos;
+    q << q_vel;
+    plant.SetPositionsAndVelocities(&plant_context, q);
+
+    // Create drake::trajectories::Trajectory from moveit trajectory
+    auto input_trajectory = getPiecewisePolynomial(*res.trajectory, joint_model_group);
+    // Run toppra
+    const auto grid_points = Toppra::CalcGridPoints(input_trajectory, CalcGridPointsOptions());
+    auto toppra = Toppra(input_trajectory, plant, grid_points);
+
+    /////////////////////////////////////////////////////////////////////////
+    // Read joint bounds from robot model (TODO(sjahr): Expose in MoveIt2) //
```

Comment:
@kamiradi FYI:This is how you can read the joint limits

Reply:
Might be good to put this in a utility function to share between TOPPRA and KTOpt (and whatever else comes next)

Reply:
I'll add that to moveit

Reply:
Cool! Might be useful in MoveIt Servo as well since I recall having to write similar code there.